google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreBundle.java

```diff
@@ -0,0 +1,228 @@
+      BundledQuery query = querySnap.getQuery().toBundledQuery();
+      NamedQuery namedQuery =
+          NamedQuery.newBuilder()
+              .setName(queryName)
+              .setReadTime(querySnap.getReadTime().toProto())
+              .setBundledQuery(query)
+              .build();
+      namedQueries.put(queryName, namedQuery);
+
+      for (QueryDocumentSnapshot snapshot : querySnap.getDocuments()) {
+        add(snapshot, Optional.of(queryName));
+      }
+
+      if (querySnap.getReadTime().compareTo(latestReadTime) > 0) {
+        latestReadTime = querySnap.getReadTime();
+      }
+
+      return this;
+    }
+
+    public FirestoreBundle build() {
+      StringBuilder buffer = new StringBuilder();
+
+      for (NamedQuery namedQuery : namedQueries.values()) {
+        buffer.append(
+            elementToLengthPrefixedStringBuilder(
+                BundleElement.newBuilder().setNamedQuery(namedQuery).build()));
+      }
+
+      for (BundledDocument bundledDocument : documents.values()) {
+        buffer.append(
+            elementToLengthPrefixedStringBuilder(
+                BundleElement.newBuilder()
+                    .setDocumentMetadata(bundledDocument.getMetadata())
+                    .build()));
+        if (bundledDocument.getDocument() != null) {
+          buffer.append(
+              elementToLengthPrefixedStringBuilder(
+                  BundleElement.newBuilder().setDocument(bundledDocument.getDocument()).build()));
+        }
+      }
+
+      BundleMetadata metadata =
+          BundleMetadata.newBuilder()
+              .setId(id)
+              .setCreateTime(latestReadTime.toProto())
+              .setVersion(BUNDLE_SCHEMA_VERSION)
+              .setTotalDocuments(documents.size())
+              .setTotalBytes(buffer.toString().getBytes().length)
+              .build();
+      BundleElement element = BundleElement.newBuilder().setMetadata(metadata).build();
+      buffer.insert(0, elementToLengthPrefixedStringBuilder(element));
+
+      return new FirestoreBundle(buffer.toString().getBytes(StandardCharsets.UTF_8));
+    }
+
+    private StringBuilder elementToLengthPrefixedStringBuilder(BundleElement element) {
+      String elementJson = null;
+      try {
+        elementJson = printer.print(element);
+      } catch (InvalidProtocolBufferException e) {
+        throw new RuntimeException(e);
+      }
+      return new StringBuilder().append(elementJson.getBytes().length).append(elementJson);
+    }
+  }
+
+  private FirestoreBundle(byte[] data) {
+    bundleData = data;
+  }
+
+  /** Returns the bundle content as a readonly {@link ByteBuffer}. */
+  public ByteBuffer toByteBuffer() {
+    return ByteBuffer.wrap(bundleData).asReadOnlyBuffer();
+  }
+}
+
+/**
+ * Convenient class to hold both the metadata and the actual content of a document to be bundled.
+ */
+class BundledDocument {
+  private BundledDocumentMetadata metadata;
+  private Document document;
```

Comment:
This class appears to be write-once, can we mark these fields as `final`?

Reply:
Done, metadata is set above though.